controllers/runnerpool_controller.go

```diff
@@ -0,0 +1,229 @@
+
+	organizationName string
+}
+
+// NewRunnerPoolReconciler creates RunnerPoolReconciler
+func NewRunnerPoolReconciler(
+	client client.Client,
+	log logr.Logger,
+	scheme *runtime.Scheme,
+	organizationName string,
+) *RunnerPoolReconciler {
+	return &RunnerPoolReconciler{
+		Client:           client,
+		Log:              log,
+		Scheme:           scheme,
+		organizationName: organizationName,
+	}
+}
+
+//+kubebuilder:rbac:groups=actions.cybozu.com,resources=runnerpools,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=actions.cybozu.com,resources=runnerpools/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *RunnerPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log.Info("start reconciliation loop", "name", req.NamespacedName)
+
+	rp := &actionsv1alpha1.RunnerPool{}
+	if err := r.Get(ctx, req.NamespacedName, rp); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("runnerpool is not found")
+			return ctrl.Result{}, nil
+		}
+		r.Log.Error(err, "unable to get RunnerPool", "name", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	if rp.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(rp, runnerPoolFinalizer) {
+			err := r.addFinalizer(ctx, rp)
+			if err != nil {
+				r.Log.Error(err, "failed to add finalizer", "name", req.NamespacedName)
+				return ctrl.Result{}, err
+			}
+			// Result does not change even if not requeued here.
+			// This just breaks down one reconciliation loop into small steps for simplicity.
+			r.Log.Info("added finalizer", "name", req.NamespacedName)
+			return ctrl.Result{Requeue: true}, nil
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(rp, runnerPoolFinalizer) {
+			err := r.cleanUpOwnedResources(ctx, req.NamespacedName)
+			if err != nil {
+				r.Log.Error(err, "failed to clean up deployment", "name", req.NamespacedName)
+				return ctrl.Result{}, err
+			}
+
+			err = r.removeFinalizer(ctx, rp)
+			if err != nil {
+				r.Log.Error(err, "failed to remove finalizer", "name", req.NamespacedName)
+				return ctrl.Result{}, err
+			}
+			r.Log.Info("removed finalizer", "name", req.NamespacedName)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	d := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+		},
+	}
+	op, err := ctrl.CreateOrUpdate(ctx, r.Client, d, func() error {
+		return r.updateDeploymentWithRunnerPool(rp, d)
+	})
+	if err != nil {
+		r.Log.Error(err, "unable to create-or-update Deployment", "name", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+	r.Log.Info("completed create-or-update successfully with status "+string(op), "name", req.NamespacedName)
+
+	rp.Status.Bound = true
```

Comment:
Is it not necessary to set `bound` to false when updating deployment fails?

Reply:
It can be set as false, but the implicit initial value of `Bound` is false, so I dare not to do that.